### PR TITLE
Add PEP 518 build; annotate types; drop Python < 3.6 support

### PR DESCRIPTION
**Pull request by djds**
- [ci] Format asciinema workflow; add Python 3.10
- [test] Shellcheck tests/*.sh
- [build] Refactor Makefile
- Convert setup.py to setup.cfg, pyproject.toml
- [test] Fix imports in tests/config_test.py
- [doc] Update README
- [container] Update Dockerfile for PEP 518 build
- [test] Update test distro Dockerfiles
- [test] Convert unittests runner nose => pytest
- [ci] Run container integration tests to CI
- [ci] Add dependabot config
- [style] Add pre-commit config
- [style] Format package code
- [style] Ignore line break before binary operator
- [style] Replace calls to format() with f-strings
- [style] Disable pylint invalid-name
- Rename asciinema/pty.py => pty_.py
- [typing] Annotate asciinema.asciicast
- [typing] Annotate asciinema.async_worker
- [test] Update test for asciinema.pty_
- [test] Drop support for Python 3.6; upgrade CentOS
- [build] Add clean, clean.all targets to Makefile
- [typing] Annotate ascinnema.urllib_http_adapter
- [typing] Annotate asciinema.api
- [typing] Annotate asciinema.config
- [typing] Annotate asciinema.term
- [typing] Annotate asciinema.pty_
- [typing] Annotate asciinema.__main__
- [typing] Annotate asciinema.player
- [typing] Annotate asciinema.recorder
- [test] Configure pylint
- [test] Format v2 test
- [test] Format test_helper
- Refactor asciinema.notifier
- [typing] Annotate asciinema.commands.auth
- [build] Run build target in VIRTUAL_ENV
- [typing] Annotate asciinema.commands.upload
- [typing] Annotate asciinema.commands.record
- [typing] Annotate asciinema.__init__
- [typing] Annotate asciinema.commands.cat
- [typing] Annotate asciinema.commands.command
- [typing] Annotate asciinema.commands.play
- [typing] Add asciinema/py.typed
- Revert "[test] Drop support for Python 3.6; upgrade CentOS"
- Support Python 3.6; remove annotations from __future__
- v2.2.0: Require Python 3.6+

### CODE REVIEW

**Comment by djds** [comment]
@sickill I know this PR is massive, but let me know if you want an explanation for anything. Also the docker tests will pass if run by someone who has rw access to this repo, for instance on a different branch, since they use CI vars for authentication"
```yaml
    container:
      image: "ghcr.io/${{ github.repository }}:${{ matrix.distros }}"
      credentials:
        username: "${{ github.actor }}"
        password: "${{ secrets.GITHUB_TOKEN }}"
```

**Comment by sickill** [comment]
Oh wow, that is massive! You've put a lot of effort into that. Thanks! I'll try to find some time in the coming days to look at it and will let you know.

**Comment by sickill** [comment]
Do you think it would be possible to split this PR into multiple smaller ones, each tackling separate aspect? Looking at all of the changes (which are super valuable!) altogether requires more effort, and that's why I've been postponing looking at this...

If it would be not too hard to let's say cherry pick the commits and group them in separate PRs for:

- packaging
- CI
- test refactor
- typing
- style

then it would be lovely.

I expect that it may not be easy in some cases, where follow up commits for a different aspect modify lines that were changed earlier for another reason, and untangling this would require more work. If that's the case then no worries, I'll try to review it all here.

Let me know what you think. Thanks!

**Comment by djds** [comment]
I'll give it a shot this week. As you suspect, some of it might be difficult to untangle, but I bet I can break it up a bit.

**Comment by stratakis** [comment]
I see that my PR #455 might conflict with some aspects of this PR. Feel free to use the commits from #455 or just utilize the code to enhance your PR here.

Noting that #455 restructures the tests directory in order to use unittest instead of nose, but pytest might be a better alternative as to not change the directory structure.

File: asciinema/commands/cat.py
Comment (sickill) [inline asciinema/commands/cat.py:17]: The reason we open /dev/tty here is to get its file descriptor and turn it into raw mode, then _read_ and ignore any terminal answerbacks, so user doesn't get garbage input in their shell prompt. Therefore open mode could (and probably should) be left as read-only.

File: setup.cfg
Comment (sickill) [inline setup.cfg]: Awesome. This should have been cleaned up a while ago.

**Review by sickill — CHANGES_REQUESTED** [review]
First let me say big thank you! There's too many individual things that make the codebase much better so I'm not going to comment on them all. This is amazing. ❤️ 🏆 

By reviewing this I learned a lot, especially about modern Python coding standards and tooling (I'm not a Python dev at $dailyjob).

There's just a few tiny things that got lost, once we have them sorted out I'm happy to merge this. This probably doesn't justify a new relase though, as it doesn't bring any new functionality nor bugfixes.

Also, could you explain that thing with Github Actions failing - what should I do to make the image building steps succeed?

File: asciinema/urllib_http_adapter.py
Comment (djds) [reply asciinema/urllib_http_adapter.py]: This should be fixed in 0ef315fcac2aff98fd7d61e9008995d1fbeb5400

**Comment by djds** [comment]
> Also, could you explain that thing with Github Actions failing - what should I do to make the image building steps succeed?

This should be fixed once things get merged as my fork doesn't have push access to the `ghcr.io/asciinema/asciinema/${image}:${tag}` container registry: https://github.com/asciinema/asciinema/runs/5174851854?check_suite_focus=true#step:4:220

**Comment by djds** [comment]
@sickill thanks for the detailed review. I think I have addressed everything: let me know if I missed anything.

File: asciinema/recorder.py
Comment (djds) [reply asciinema/recorder.py]: Sorry I phrased things poorly there. `assert` statements should _not_ be relied upon to impact program behavior or as control flow or as a security check because they _may_ be removed depending on the compiler options as you point out. I doubt that these assertions are costly, but honestly I'm not an expert on the internals of CPython. I think the best approach may be to leave them for now as indicators to be refactored out later as certain behaviors are changed to become more deterministic or handle a specific error case.

Comment (sickill) [reply asciinema/recorder.py]: Got it. Sounds good 👍

File: asciinema/commands/command.py
Comment (djds) [reply asciinema/commands/command.py]: :facepalm:

**Comment by sickill** [comment]
There seem to be some conflict, I'll try to resolve this during merge.

**Pull request merged**
